File: app/src/main/java/net/osdn/gokigen/gr2control/Gr2ControlMain.java
package net.osdn.gokigen.gr2control;

import android.Manifest;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentTransaction;
import androidx.preference.PreferenceManager;

import android.os.Bundle;
import android.util.Log;
import android.view.KeyEvent;
import android.view.WindowManager;

import net.osdn.gokigen.gr2control.camera.CameraInterfaceProvider;
import net.osdn.gokigen.gr2control.camera.ICameraConnection;
import net.osdn.gokigen.gr2control.camera.IInterfaceProvider;
import net.osdn.gokigen.gr2control.liveview.LiveViewFragment;
import net.osdn.gokigen.gr2control.preference.IPreferencePropertyAccessor;
import net.osdn.gokigen.gr2control.scene.CameraSceneUpdater;

/**
 *
 *
 */
public class Gr2ControlMain extends AppCompatActivity
{
    private final String TAG = toString();
    private IInterfaceProvider interfaceProvider = null;
    private CameraSceneUpdater scenceUpdater = null;
    private  LiveViewFragment liveViewFragment = null;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        // 画面表示の準備
        setContentView(R.layout.activity_gr2_control_main);
        try
        {
            ActionBar bar = getSupportActionBar();
            if (bar != null)
            {
                // タイトルバーは表示しない
                bar.hide();
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        // 外部メモリアクセス権のオプトイン
        final int REQUEST_NEED_PERMISSIONS = 1010;
        if ((ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) ||
                (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_NETWORK_STATE) != PackageManager.PERMISSION_GRANTED) ||
                (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_WIFI_STATE) != PackageManager.PERMISSION_GRANTED) ||
                (ContextCompat.checkSelfPermission(this, Manifest.permission.VIBRATE) != PackageManager.PERMISSION_GRANTED) ||
                (ContextCompat.checkSelfPermission(this, Manifest.permission.INTERNET) != PackageManager.PERMISSION_GRANTED)) {
            ActivityCompat.requestPermissions(this,
                    new String[]{
                            Manifest.permission.WRITE_EXTERNAL_STORAGE,
                            Manifest.permission.ACCESS_NETWORK_STATE,
                            Manifest.permission.ACCESS_WIFI_STATE,
                            Manifest.permission.VIBRATE,
                            Manifest.permission.INTERNET,
                    },
                    REQUEST_NEED_PERMISSIONS);
        }

        initializeClass();
        onReadyClass();
    }

    /**
     *
     *
     */
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String  permissions[], @NonNull int[] grantResults)
    {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        onReadyClass();
    }

    /**
     *
     */
    @Override
    protected void onPause()
    {
        super.onPause();
        try
        {
            ICameraConnection connection = interfaceProvider.getCameraConnection();
            if (connection != null)
            {
                connection.stopWatchWifiStatus(this);
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    /**
     * クラスの初期化 (instantiate)
     *
     */
    private void initializeClass()
    {
        try
        {
            scenceUpdater = CameraSceneUpdater.newInstance(this);
            interfaceProvider = CameraInterfaceProvider.newInstance(this, scenceUpdater);

            if (liveViewFragment == null)
            {
                liveViewFragment = LiveViewFragment.newInstance(scenceUpdater, interfaceProvider);
                scenceUpdater.registerInterface(liveViewFragment, interfaceProvider);
            }
            liveViewFragment.setRetainInstance(true);
            FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
            transaction.replace(R.id.fragment1, liveViewFragment);
            transaction.commitAllowingStateLoss();
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    /**
     *    初期化終了時の処理 (カメラへの自動接続)
     */
    private void onReadyClass()
    {
        try
        {
            // カメラに自動接続するかどうか確認
            SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
            boolean isAutoConnectCamera = preferences.getBoolean(IPreferencePropertyAccessor.AUTO_CONNECT_TO_CAMERA, true);
            Log.v(TAG, "isAutoConnectCamera() : " + isAutoConnectCamera);

            // カメラに接続する
            if (isAutoConnectCamera)
            {
                // 自動接続の指示があったとき
                scenceUpdater.changeCameraConnection();
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event)
    {
        Log.v(TAG, "onKeyDown()" + " " + keyCode);
        try
        {
            if ((event.getAction() == KeyEvent.ACTION_DOWN)&&
                    ((keyCode == KeyEvent.KEYCODE_VOLUME_UP)||(keyCode == KeyEvent.KEYCODE_CAMERA)))
            {
                if (liveViewFragment != null)
                {
                    return (liveViewFragment.handleKeyDown(keyCode, event));
                }
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        return (super.onKeyDown(keyCode, event));
    }
}

File: app/src/main/java/net/osdn/gokigen/gr2control/Gr2ControlMain.kt
package net.osdn.gokigen.gr2control

import android.Manifest.permission
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.KeyEvent
import android.view.View
import android.view.WindowManager
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updatePadding
import androidx.preference.PreferenceManager
import net.osdn.gokigen.gr2control.camera.CameraInterfaceProvider
import net.osdn.gokigen.gr2control.camera.IInterfaceProvider
import net.osdn.gokigen.gr2control.liveview.LiveViewFragment
import net.osdn.gokigen.gr2control.preference.IPreferencePropertyAccessor
import net.osdn.gokigen.gr2control.scene.CameraSceneUpdater

class Gr2ControlMain : AppCompatActivity() {

    private lateinit var interfaceProvider: IInterfaceProvider
    private lateinit var sceneUpdater: CameraSceneUpdater
    private lateinit var liveViewFragment: LiveViewFragment

    // 実行中の OS バージョンに応じて要求する権限一覧を取得
    private val requiredPermissions: Array<String>
        get()
        {
            val permissions = mutableListOf<String>()

            // 通常権限
            permissions.add(permission.ACCESS_NETWORK_STATE)
            permissions.add(permission.ACCESS_WIFI_STATE)
            permissions.add(permission.INTERNET)
            permissions.add(permission.VIBRATE)

            // API 28 (Android 9.0) 以下のみストレージ書き込み権限を要求
            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P)
            {
                permissions.add(permission.WRITE_EXTERNAL_STORAGE)
            }

            // API 33 (Android 13) 以上で Wi-Fi 機器検出権限を追加
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
            {
                permissions.add(permission.NEARBY_WIFI_DEVICES)
            }

            // API 37 (Android 17) 以上向けローカルネットワーク権限
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.CINNAMON_BUN)
            {
                permissions.add("android.permission.ACCESS_LOCAL_NETWORK")
            }

            return permissions.toTypedArray()
        }

    private val requestPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { _ ->
        if (allPermissionsGranted()) {
            Log.v(TAG, "Permissions granted via launcher")
            setupAndConnect()
        } else {
            Log.v(TAG, "----- APPLICATION LAUNCH ABORTED (Permission Rejected) -----")
            Toast.makeText(this, getString(R.string.permission_not_granted), Toast.LENGTH_SHORT).show()
            finish()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_gr2_control_main)

        supportActionBar?.hide()
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        val baseLayout = findViewById<View>(R.id.base_layout)
        if (baseLayout != null) {
            setupWindowInset(baseLayout)
        }

        // 権限の確認とリクエスト
        Log.v(TAG, " ----- SET PERMISSIONS -----")
        if (allPermissionsGranted()) {
            setupAndConnect()
        } else {
            requestPermissionLauncher.launch(requiredPermissions)
        }
    }

    private fun setupAndConnect() {
        initializeClass()
        onReadyClass()
    }

    private fun setupWindowInset(view: View) {
        ViewCompat.setOnApplyWindowInsetsListener(view) { v, insets ->
            val bars = insets.getInsets(
                WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout()
            )
            v.updatePadding(
                left = bars.left,
                top = bars.top,
                right = bars.right,
                bottom = bars.bottom,
            )
            WindowInsetsCompat.CONSUMED
        }
    }

    private fun allPermissionsGranted(): Boolean {
        for (param in requiredPermissions) {
            if (ContextCompat.checkSelfPermission(this, param) != PackageManager.PERMISSION_GRANTED) {
                Log.v(TAG, " Permission denied: $param (SDK: ${Build.VERSION.SDK_INT})")
                return false
            }
        }
        return true
    }

    override fun onPause() {
        super.onPause()
        try {
            if (::interfaceProvider.isInitialized) {
                interfaceProvider.cameraConnection?.stopWatchWifiStatus(this)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error in onPause", e)
        }
    }

    // クラスの初期化
    private fun initializeClass() {
        try {
            if (!::sceneUpdater.isInitialized) {
                sceneUpdater = CameraSceneUpdater.newInstance(this)
            }
            if (!::interfaceProvider.isInitialized) {
                interfaceProvider = CameraInterfaceProvider.newInstance(this, sceneUpdater)
            }
            if (!::liveViewFragment.isInitialized) {
                liveViewFragment = LiveViewFragment.newInstance(sceneUpdater, interfaceProvider)
            }

            sceneUpdater.registerInterface(liveViewFragment, interfaceProvider)

            // Fragmentの重複生成帽子
            if (supportFragmentManager.findFragmentById(R.id.fragment1) == null) {
                supportFragmentManager.beginTransaction()
                    .replace(R.id.fragment1, liveViewFragment)
                    .commitAllowingStateLoss()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error in initializeClass", e)
        }
    }

    // 初期化終了時の処理 (カメラへの自動接続)
    private fun onReadyClass() {
        try {
            val preferences = PreferenceManager.getDefaultSharedPreferences(this)
            val isAutoConnectCamera = preferences.getBoolean(
                IPreferencePropertyAccessor.AUTO_CONNECT_TO_CAMERA,
                true
            )
            Log.v(TAG, "isAutoConnectCamera() : $isAutoConnectCamera")

            if (isAutoConnectCamera && ::sceneUpdater.isInitialized) {
                sceneUpdater.changeCameraConnection()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error in onReadyClass", e)
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        Log.v(TAG, "onKeyDown() $keyCode")
        if (event.action == KeyEvent.ACTION_DOWN &&
            (keyCode == KeyEvent.KEYCODE_VOLUME_UP || keyCode == KeyEvent.KEYCODE_CAMERA)
        ) {
            if (::liveViewFragment.isInitialized) {
                return liveViewFragment.handleKeyDown(keyCode, event)
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    companion object {
        private val TAG = Gr2ControlMain::class.java.simpleName
    }
}
